lib/reversetunnel/seek/seek.go

```diff
@@ -239,37 +300,55 @@ func (s *GroupHandle) Gossip() chan<- string {
 // proxyGroup manages all proxy seekers for a group.
 type proxyGroup struct {
 	sync.Mutex
-	id       Key
+	*log.Entry
+	key      Key
 	conf     Config
 	states   map[string]seeker
 	proxyC   <-chan string
-	seekC    chan<- Key
+	grantC   chan<- *Lease
+	releaseC chan struct{}
 	statC    chan Status
+	notifyC  chan struct{}
 	lastStat *Status
+	ids      counter
 }
 
 // run is the "main loop" for the seek process.
-func (p *proxyGroup) run(ctx context.Context) {
-	const logInterval int = 8
+func (p *proxyGroup) run(ctx context.Context, handle GroupHandle, retry utils.Retry) {
+	const logInterval uint64 = 8
+	// ticker is the primary ticker used to schedule
+	// group-level "ticks".
 	ticker := time.NewTicker(p.conf.TickRate)
 	defer ticker.Stop()
-	// supply initial status & seek notification.
-	p.notifyStatus(p.Tick())
-	p.notifyShouldSeek()
-	var ticks int
+	// backoff is a seek notification backoff channel.  The main tick
+	// loop applies backoff after generating seek notifications.
+	var backoff <-chan time.Time
+	// nextLease may hold the next lease to be granted if we were
+	// interrupted while blocking on the lease grant channel.
+	var nextLease *Lease
+	// activeLeases tracks the number of leases that have been granted.
+	activeLeases := 0
+	var ticks, lastLog uint64
+	// start the loop in a notified state
+	p.notifyGroup()
 	for {
 		select {
+		case <-ctx.Done():
+			// this group is closing.
+			return
 		case <-ticker.C:
-			stat := p.Tick()
-			p.notifyStatus(stat)
-			if stat.ShouldSeek() {
-				p.notifyShouldSeek()
-			}
+			// when ticker fires, we should recalculate state.
 			ticks++
-			if ticks%logInterval == 0 {
-				log.Debugf("SeekStates(states=%+v,id=%s)", p.GetStates(), p.id)
-			}
+			p.notifyGroup()
+		case <-p.releaseC:
+			// a lease has been relenquished, update activeLeases count and
+			// trigger recalculation of state.
+			activeLeases--
+			p.notifyGroup()
```

Comment:
this seems like it notifies itself via channel send? I can be wrong, but on the first sight this logic can give up the lease, but notify has a `default` clause  that can trigger in case if any blocked channel and then new lease will never be issued.